### ISSUE TITLE
jax.checkpoint does not reduce memory footprint

### ISSUE THREAD

**Issue by dionhaefner** [google/jax#19822]
### Description

I was doing experiments with `jax.checkpoint` on a simple neural network. I noticed that use of `jax.checkpoint` never actually managed to decrease memory consumption for me (testing on CPU). 

I use these functions for testing:

```python
num_layers = 10

@jax.jit
def foobar(x):
    # fake feedforward network
    key = jax.random.PRNGKey(0)
    for _ in range(num_layers):
        key, subkey = jax.random.split(key)
        random_matrix = jax.random.normal(subkey, shape=(x.shape[0], x.shape[0]))
        x = jnp.dot(random_matrix, x) + 1
        x = jax.nn.relu(x)
    return x.sum()

foobar_checkpointed = jax.checkpoint(foobar)
grad_fn = jax.value_and_grad(foobar)
grad_fn_checkpointed = jax.value_and_grad(foobar_checkpointed)
```

Example memory usage:

<img width="563" alt="Screenshot 2024-02-15 at 11 56 50" src="https://github.com/google/jax/assets/11994217/f7f2dd64-3da5-4b90-966f-66f92b80a9bf">

I.e., the checkpointed function is slower as expected, but has the same peak memory usage as the non-checkpointed version.

Here's the notebook I used to profile this:

[jaxbug.ipynb.zip](https://github.com/google/jax/files/14295845/jaxbug.ipynb.zip)

Any pointers as to what might be going on?

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.4.24
jaxlib: 0.4.24
numpy:  1.26.1
python: 3.10.13 (main, Aug 24 2023, 12:59:26) [Clang 15.0.0 (clang-1500.0.40.1)]
jax.devices (1 total, 1 local): [CpuDevice(id=0)]
process_count: 1
```

**Comment by nireekshak** [comment]
Hi @dionhaefner, The `jax.checkpoint` is intended to reduce memory usage, the small model size and underlying optimizations from XLA and JIT compilation might limit its effectiveness. Even with the large memory used in `grad_fn` and `grad_fn_checkpointed` is because of the gradients. Checkpointing the forward pass doesn't reduce memory usage during backpropagation, as gradients still require memory for intermediate values and since checkpointing is most effective for larger models with significant intermediate activations, its impact may be limited in smaller models.

Thank you.

**Comment by dionhaefner** [comment]
Thanks @nireekshak. I think this statement is correct:

> underlying optimizations from XLA and JIT compilation might limit its effectiveness

But I don't see the argument regarding small model size. Checkpointing is alleviating the need to store intermediate values for the backward pass. Why would it matter how large the model is? (And feel free to change the depth or size of those matrices; when I tried it that didn't make a difference.)

**Comment by AntoinePlumerault** [comment]
Hello I think the issue is that you are checkpointing your entire forward function. So during the forward pass no intermediate result is saved. But during the backward pass JAX has to run your entire forward pass again and save all intermediate activations to compute the gradient using backpropagation. You are suppose to encapsulate parts of the forward with jax.checkpoint to reduce memory usage.

**Comment by dionhaefner** [comment]
Is that so? It was my understanding that `jax.checkpoint` prevents *all* storage of primals. That is, the function:

```
f := (inp) -> A -> B -> C -> D -> (out)
```

has this backwards pass:

```
f_grad := (Cp -> D') -> (Bp -> C') -> (Ap -> B') -> (inp -> A')
```

(where `Xp` are saved primals of operator `X`)

Under checkpointing, it would compute this:

```
f_chkp := checkpoint(f)
f_chkp_grad := (A -> B -> C -> D') -> (A -> B -> C') -> (A -> B') -> (inp -> A')
```

I.e., no primals are saved ever. What you suggest sounds more like this:

```
f_chkp_grad := (A -> B -> C -> D') -> (Bp -> C') -> (Ap -> B') -> (inp -> A')
```

which would indeed not reduce memory usage.
